linera-indexer/src/operations.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module defines the operations indexer plugin.
+
+use crate::types::IndexerError;
+use async_graphql::{
+    EmptyMutation, EmptySubscription, Object, OneofObject, OutputType, Schema, SimpleObject,
+};
+use linera_base::{
+    crypto::{BcsHashable, CryptoHash},
+    data_types::BlockHeight,
+    identifiers::ChainId,
+};
+use linera_chain::data_types::HashedValue;
+use linera_execution::Operation;
+use linera_views::{
+    common::Context,
+    map_view::MapView,
+    views::{RootView, View, ViewError},
+};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use tokio::sync::Mutex;
+use tracing::info;
+
+#[derive(Deserialize, Serialize, Clone, SimpleObject, Debug)]
+pub struct BlockOperation {
+    block: CryptoHash,
+    index: usize,
+    content: linera_execution::Operation,
+}
+
+impl BcsHashable for BlockOperation {}
+
+#[derive(Deserialize, Serialize, Clone, SimpleObject, Debug)]
+pub struct ChainOperation<H: Sync + Send + OutputType> {
+    operation: BlockOperation,
+    height: BlockHeight,
+    previous_operation: Option<CryptoHash>,
+    operation_index: u64,
+    hash: H,
+}
+
+#[derive(RootView)]
+pub struct OperationsPluginInternal<C> {
+    last: MapView<C, ChainId, (CryptoHash, BlockHeight, usize)>,
+    count: MapView<C, ChainId, u64>,
+    /// ChainOperation MapView indexed by their hash
+    operations: MapView<C, CryptoHash, ChainOperation<bool>>,
+}
+
+#[derive(Clone)]
+pub struct OperationsPlugin<C>(Arc<Mutex<OperationsPluginInternal<C>>>);
+
+impl<C> OperationsPlugin<C>
+where
+    C: Context + Send + Sync + 'static + Clone,
+    ViewError: From<C::Error>,
+{
+    /// Registers an operation and update count and last entries for this chain ID
+    async fn register_operation(
+        &self,
+        content: Operation,
+        block: CryptoHash,
+        height: BlockHeight,
+        index: usize,
+        chain_id: ChainId,
+    ) -> Result<(), IndexerError> {
+        let mut plugin = self.0.lock().await;
+        let last_operation = plugin.last.get(&chain_id).await?;
+        match last_operation {
+            Some((_last_hash, last_height, last_index))
+                if ((last_height > height) || (last_height == height) && (last_index > index)) =>
+            {
+                Ok(())
+            }
+            last_operation => {
+                let previous_operation = last_operation.map(|last| last.0);
+                let operation = BlockOperation {
+                    block,
+                    index,
+                    content,
+                };
+                let key = CryptoHash::new(&operation);
```

Comment:
Why not use (chain ID, block height, operation index) as the key?

Reply:
It would also work since this hash is not used for security